Linting broken for script tag (auto-fix mangles file)

### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.

### What version of ESLint are you using?

8.36.0

### What version of `eslint-plugin-astro` are you using?

0.26.0

### What did you do?

<details>
<summary>Configuration</summary>

```
module.exports = {
  extends: [
    'plugin:astro/recommended',
    'standard-with-typescript'
  ],
  parserOptions: {
    project: './tsconfig.json'
  }
}

```
</details>

```astro
<!DOCTYPE html>
<html lang="de">
  <head>
    <script>
      alert('test')
    </script>
  </head>
  <body>
  </body>
</html>
```


### What did you expect to happen?

I expected no errors and --fix to correctly format the file

### What actually happened?

ESLint shows errors:
```
  6:4  error  Trailing spaces not allowed                                no-trailing-spaces
  6:4  error  Too many blank lines at the end of file. Max of 0 allowed  no-multiple-empty-lines
  6:5  error  Newline required at end of file but not found              eol-last
```

And when running `eslint --fix` on the file, the result is completely broken:
```
<!DOCTYPE html>
<html lang="de">
  <head>
    <script>
      alert('test')
   <!DOCTYPE html>
<html lang="de">
  <head>
    <script>
      alert('test')

<!DOCTYPE html>
<html lang="de">
  <head>
    <script>
      alert('test')
    </script>
  </head>
  <body>
  </body>
</html>
```

### Link to Minimal Reproducible Example

https://ota-meshi.github.io/eslint-plugin-astro/playground/#eJyVWcuuHLcR/ZWbu3ECmBCcpaFk42QdL7wJoA2nm9NDq5tskex52PC/51Tx3dMjORAEDauq+ajHqUPq9/fBjur9x/ePf/nXf3765b8///vtEpb5n5/MR/r3bZZm+sen91F9eofs7e3jRcmRf+G3H5xeQxq9vclZufDX74Ly4bu/JZMPjc3HD+Xjjyc7PpIw/YQ2Lvz+/bvbZuXff/z9Xfrg7AdjxWDNedZDEF4FMWqnhqCvZPP376vRqFYoZFCjYJkYpLFGD3Le3PwN07MKwwWrBGXCN0yd8na+qt6K9kUHqJt71gd1D6/0m9k8Vhi8h+kMC+sODUZ11kaJq3ReaCN8eMztTNjqWTkxzBITzdofr5es7OlXqBrjZxO/zvqFhVeLboZXOWvs3y6r7nb0q78L+cMPjw9yjh441JnhYp2Qy0lPm938twwv0ovnYO2ttBe8q0MTp6WQ7JhR+UGZUZrA8wZ50mZU95dfrc6u/qva8FjVawtnXziIlJvx27paR+mGPFhwwuOZtmDJ2bMK6vUpB9TMZ6GuNAsOd1Xis3qk8aE9fIrtsR8QcjtozvtZntR8ZE8Frc30rXhwYZANwcmRgT47uajofh2O3aOXCaU3bjFSyKZDI+zBxbCK5EgvznbYDo/L59ofNvng2P54+4saETp2glyDtubQCHmt/o9YoOblMCgUHuxeGVDKXPQ4KiOsiQeVp2P/kTmS5qUzCOpQ4ew9hPRryQfT1tHJVAQrCG8bzatUN0+WeY4is+bV0i++DP2m/vzSXyt4Y5ucoylfbcoHGfTwp4+R4DXISdgrfrzaLck5sZz6sgHJR/F1CGL7kvhft/WDXQ8XzQ6Bo8RqvS7dA6UBy42aEy2iYos6Y5S6TMr8SYVAh1Jhc1FCyecfZqDdLJrq4K6GLTc5avASwlGh0zhZvUYK7j34q6cyFQGfdEoYNYnflLNVbsYnU2z5SEYBJXtddh01jH9OTSkfmACctmlSro4T6FYJfAFnT74TxJ0vajkp12vUjMPqcydDkbdGAG6gEbslS9WCxtINxHCRlGW54/fKVVIUytkQz94NaLFOipNFykhDK4VeUxo8JGc5z+Hi7DZdimhDOPsZp9me5LwT6oWScS80kS+wo9cidRhvSAFxu2jwyFUO5fSzhTPtWRAZ076JGdJpTm3ohTsWfUfh8GzIJKQIErwojboJ/1hOCfEjPNgh4BxIR73gX/ADuZaNgDXBWfNcZmDjPEB+B0v9X5w2PQNJixmKl9NbVQHOnraVZeB+552r/FEx+Ysc7Y2oaHCgxTieQQOt06A8PKWkkzWrwgWE6KRQrqqpXqaWYJV1gIAonnLZcIAZdLPqnJJwcu4wLPLyrARIKVzy2ElRn7ItryS23CPhWEylAcVT1TPH5bbsd0JivEVEFBnpcJLDZwZSZWqiZCXqZ7jshX0obzpEi4Su4qHVHIkU9WrtjYx7j8lKYbXRUexZTgPawx7pblJTC0DO2pLbBXSArdKh8d+Ryd734FPQtZ+w4kEspuIIbQxS4wg2Ka2N5P6GfG6mbhK1A+J9glEiJpwtCQRcmRmVcJ96xOuH68K32FGfNYUQJ3+G1yZ/Ot+kEK9OX3n6I+CkgKTkjXegLnAy4DSD2NYRE6Rux+QJB1ulTtMgaigaunA2F6fTbBFBboZjAfUB1TQX8EXOaNS4/o3p4VJ5Ud5puNjR8xapyFKvJHJ+1wmyKQJgV8QNGkc30vohOZd/4ZwSJVibQCtAlSTAzlKESi6N2AaEPtQCVF/oT2zZhN2EGGKhKskxLGJfh9VRE+CffHTg2GmTbhREBHRcS9NN1TzKlRHjWZkplRtGS6lOIEB4TuHZTnRxT1jInAoBgoVNSy7yHhMFmA4FEChtkxS4s6elaEQQ5vtR/ATnK94hFRR8A0hlXb9h19Yhkb2GHReczMnROI3qEOlToIHeSMoggshzbZ50uOna+Gk/Fa3pKWTz1O44nVv0sBWXiclos5XxqK89q2EO0td8pA2dV4o08VsulqL5Isw2l9anrrUNgkHgSp0QaEc4wC17Sb1eEq2YreQLSOq+DY0ABtJjhFVuaPZXFJF9+E5Ot+hmW9q0YWpgNAF8rkGShZhthX/wXbEhfX2Xmi3mZe90IoLJcxUAFMm9hZhIJDl82AFdZCt9spOYkmxHDaIM2VNRf8p32B55U2YTH0TzaZpE2y/ardE4vg+1kpuT69rsNfKkvqlGGALQd1td583T344nVSZUaAzaYtvgGs0uvI1m1xQbDd19UHBaHSlxHwnyXhVUCjsHZyE19NIZ+UVT5CdF7pVfNmIglX4xO6tts3M60egbUIjyq0zBFEwQFHYSJOvYChzaFDVAaaaWghkFbvkU3dRVK9F44k8ELE8y9K2N3JNfHTqdAQs5kO4BLKtAzNBNnqUt6jQldbW6HPcmnWFO39ZqerH0F8T7IhOOWFMvY+l3208Y6guLzXXVNBaWpwt5ZAi5AbQaPmUrGFU89VYoUFJQChvMq+PyecXWhLwy8mPRhnByY+3U26KQpDlJfKtLLqA3AXszBxqPoRxbRaZ6TrGBwhldN2e8iRytRuXSdr/yLLxfI3NDln3ZLGhcfXhwuJ7de75WKipLNqPpvyLaG6Gne2Nb3CwoF2UelWsB36MKPYsirvT4k294FDOu3oyM/IyOGIXERR92lPwjtuYTbpOfVSDoKwnUa2jVHP6oyS9Au2+QVUSl82sQC9qPEwttJVhCNUyCTibb0o+C9pMkaT+JtZxw8NFZE0HEqg1BhIcKfWTyR4XQfZPJQdRIN21Phy0mfN5ONSmTqg90wnUTly4eXUOfnJBjnxNFHHNM6ZGK0ys6kp5Q22kwvlk3drKOmcX3rHTgskrjs0gPJUrS9NkUFScVbgrHer6jMLdUlUxWktgjUGWLHV6rLkGS1/hLviUnWtoTp/aDwga69wSW4R6SWFxlwtSImlcQzEZ7mnaf12eYfPfKmZQsTMrbfNZ4ueqcnP+Xie+MbTZ0ijZgSVGStvso43efIascqQM2BIwkdCB2Y47ajrg3icRAF2Gt20t5BKMfT5ombzovNfvo5H1tNHpt2mBm2VnfRcbQKNv47aDIbhq3KGQpqGa3t91N/VBFb8+tIgPwyS48Jq4ntD6rOqoXCH4o2VfxH3/8Dzamc9Q=

### Additional comments

_No response_

--- a/.changeset/modern-windows-glow.md
+++ b/.changeset/modern-windows-glow.md
@@ -1,0 +1,5 @@
+---
+"eslint-plugin-astro": patch
+---
+
+fix: compatibility with `eol-last` rule

--- a/package.json
+++ b/package.json
@@ -55,7 +55,7 @@
     "@eslint-community/eslint-utils": "^4.2.0",
     "@jridgewell/sourcemap-codec": "^1.4.14",
     "@typescript-eslint/types": "^5.25.0",
-    "astro-eslint-parser": "^0.13.0",
+    "astro-eslint-parser": "^0.13.1",
     "postcss": "^8.4.14",
     "postcss-selector-parser": "^6.0.10"
   },

--- a/src/shared/client-script/index.ts
+++ b/src/shared/client-script/index.ts
@@ -99,7 +99,7 @@ export class ClientScript {
     const indent = getIndent(lines)
     for (let index = 0; index < lines.length; index++) {
       const line = lines[index]
-      const lineIndent = Math.min(indent, line.length - 1)
+      const lineIndent = Math.min(indent, line.length)
       textLines.push(line.slice(lineIndent))
       remapLines.push(startLoc.line + index + 1)
       remapColumnOffsets.push(lineIndent)
@@ -109,26 +109,42 @@ export class ClientScript {
     const textLocs = new Locs(textLines)
 
     /** Remap loc */
-    const remapRange = (range: [number, number]): [number, number] => {
-      return [
-        this.parsed.getIndexFromLoc(
-          remapLoc(textLocs.getLocFromIndex(range[0])),
-        ),
-        this.parsed.getIndexFromLoc(
-          remapLoc(textLocs.getLocFromIndex(range[1])),
-        ),
-      ]
-    }
-
-    /** Remap loc */
-    function remapLoc(loc: { line: number; column: number }): {
+    const remapLoc = (loc: {
       line: number
       column: number
-    } {
-      return {
-        line: remapLines[loc.line - 1] ?? -1,
-        column: loc.column + (remapColumnOffsets[loc.line - 1] ?? 0),
+    }): {
+      line: number
+      column: number
+    } => {
+      const lineIndex = loc.line - 1
+      if (remapLines.length > lineIndex) {
+        return {
+          line: remapLines[lineIndex],
+          column: loc.column + remapColumnOffsets[lineIndex],
+        }
       }
+      if (remapLines.length === lineIndex) {
+        return this.parsed.getLocFromIndex(endOffset + loc.column)
+      }
+      return {
+        line: -1,
+        column: loc.column + 0,
+      }
+    }
+
+    /** Remap range */
+    const remapRange = (range: [number, number]): [number, number] | null => {
+      const startLoc = textLocs.getLocFromIndex(range[0])
+      const endLoc = textLocs.getLocFromIndex(range[1])
+      const remappedStartLoc = remapLoc(startLoc)
+      const remappedEndLoc = remapLoc(endLoc)
+      if (remappedStartLoc.line < 0 || remappedEndLoc.line < 0) {
+        return null
+      }
+      return [
+        this.parsed.getIndexFromLoc(remappedStartLoc),
+        this.parsed.getIndexFromLoc(remappedEndLoc),
+      ]
     }
 
     return {
@@ -150,11 +166,24 @@ export class ClientScript {
         }
 
         if (message.fix) {
-          message.fix.range = remapRange(message.fix.range)
+          const remappedRange = remapRange(message.fix.range)
+          if (remappedRange) {
+            message.fix.range = remappedRange
+          } else {
+            delete message.fix
+          }
         }
         if (message.suggestions) {
-          for (const suggestion of message.suggestions) {
-            suggestion.fix.range = remapRange(suggestion.fix.range)
+          for (const suggestion of [...message.suggestions]) {
+            const remappedRange = remapRange(suggestion.fix.range)
+            if (remappedRange) {
+              suggestion.fix.range = remappedRange
+            } else {
+              message.suggestions.splice(
+                message.suggestions.indexOf(suggestion),
+                1,
+              )
+            }
           }
         }
 
@@ -208,8 +237,17 @@ export class ClientScript {
 
   public remapMessages(messages: Linter.LintMessage[]): Linter.LintMessage[] {
     return messages
+      .filter((m) => !this.isIgnoreMessage(m))
       .map((m) => this.block.remapMessage(m))
       .filter((m) => m.line >= 0)
+  }
+
+  private isIgnoreMessage(message: Linter.LintMessage) {
+    if (message.ruleId === "eol-last" && message.messageId === "unexpected") {
+      // Ignore this report as it is not possible to remove the EOF of a block.
+      return true
+    }
+    return false
   }
 }
 

--- a/src/utils/transform/lines-and-columns.ts
+++ b/src/utils/transform/lines-and-columns.ts
@@ -1,6 +1,8 @@
 export class LinesAndColumns {
   private readonly lineStartIndices: number[]
 
+  private readonly code: string
+
   public constructor(code: string) {
     const len = code.length
     const lineStartIndices = [0]
@@ -16,6 +18,7 @@ export class LinesAndColumns {
         lineStartIndices.push(index + 1)
       }
     }
+    this.code = code
     this.lineStartIndices = lineStartIndices
   }
 
@@ -28,10 +31,15 @@ export class LinesAndColumns {
   }
 
   public getIndexFromLoc(loc: { line: number; column: number }): number {
-    const lineStartIndex = this.lineStartIndices[loc.line - 1]
-    const positionIndex = lineStartIndex + loc.column
-
-    return positionIndex
+    const lineIndex = loc.line - 1
+    if (this.lineStartIndices.length > lineIndex) {
+      const lineStartIndex = this.lineStartIndices[lineIndex]
+      const positionIndex = lineStartIndex + loc.column
+      return positionIndex
+    } else if (this.lineStartIndices.length === lineIndex) {
+      return this.code.length + loc.column
+    }
+    return this.code.length + loc.column
   }
 }
 

--- a/tests/src/integration/eol-last.ts
+++ b/tests/src/integration/eol-last.ts
@@ -1,0 +1,74 @@
+import { RuleTester, Linter } from "eslint"
+import { processor } from "../../../src/processor"
+
+describe("Integration test for eol-last", () => {
+  const eolLast = new Linter().getRules().get("eol-last")!
+  const tester = new RuleTester({
+    parser: require.resolve("./auto-parser"),
+    parserOptions: {
+      ecmaVersion: 2020,
+      sourceType: "module",
+    },
+  })
+  tester.run("eol-last", eolLast, {
+    valid: [
+      {
+        // @ts-expect-error -- fine name with processor
+        filename: {
+          filename: "foo.astro",
+          ...processor,
+        },
+        code: `
+        <script define:vars={{ foo: 42 }}>
+          console.log(foo)
+        </script>\n`,
+      },
+      {
+        // @ts-expect-error -- fine name with processor
+        filename: {
+          filename: "foo.astro",
+          ...processor,
+        },
+        code: `
+        <script define:vars={{ foo: 42 }}>
+          console.log(foo)
+        </script>`,
+        options: ["never"],
+      },
+      {
+        // @ts-expect-error -- fine name with processor
+        filename: {
+          filename: "foo.astro",
+          ...processor,
+        },
+        code: `
+        <script define:vars={{ foo: 42 }}>
+          console.log(foo)</script>`,
+        options: ["never"],
+      },
+    ],
+    invalid: [
+      {
+        // @ts-expect-error -- fine name with processor
+        filename: {
+          filename: "foo.astro",
+          ...processor,
+        },
+        code: `
+        <script define:vars={{ foo: 42 }}>
+          console.log(foo)</script>\n`,
+        output: `
+        <script define:vars={{ foo: 42 }}>
+          console.log(foo)
+</script>\n`,
+        errors: [
+          {
+            message: "Newline required at end of file but not found.",
+            line: 3,
+            column: 27,
+          },
+        ],
+      },
+    ],
+  })
+})

